sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,249 @@
+            CancellationToken cancellationToken = default)
+        {
+            if (TryGetLength(content, out long length) && length < _singleBlockThreshold)
+            {
+                return _client.Upload(content, blobHttpHeaders, metadata, conditions, accessTier, progressHandler, cancellationToken);
+            }
+
+            return UploadInSequence(content, blobHttpHeaders, metadata, conditions, progressHandler, accessTier, cancellationToken);
+        }
+
+        private Response<BlobContentInfo> UploadInSequence(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier,
+            CancellationToken cancellationToken)
+        {
+            List<string> blocks = new List<string>();
+
+            IAsyncEnumerator<StreamPartition> enumerator = GetBlocksAsync(content, async: false, cancellationToken).
+                GetAsyncEnumerator(cancellationToken);
+
+            while (enumerator.MoveNextAsync().EnsureCompleted())
+            {
+                using StreamPartition block = enumerator.Current;
+                string blockId = GenerateBlockId(block.AbsolutePosition);
+
+                _client.StageBlock(
+                    blockId,
+                    new MemoryStream(block.Bytes, 0, block.Length, writable: false),
+                    conditions: conditions,
+                    progressHandler: progressHandler,
+                    cancellationToken: cancellationToken);
+
+                blocks.Add(blockId);
+            }
+
+            return _client.CommitBlockList(blocks, blobHttpHeaders, metadata, conditions, accessTier, cancellationToken);
+        }
+
+        private async Task<Response<BlobContentInfo>> UploadInParallelAsync(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions blobAccessConditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier,
+            CancellationToken cancellationToken)
+        {
+            List<string> blocks = new List<string>();
+            List<Task> runningTasks = new List<Task>();
+
+            await foreach (StreamPartition block in GetBlocksAsync(content, async: true, cancellationToken))
+            {
+                string blockId = GenerateBlockId(block.AbsolutePosition);
+
+                Task task = StageBlobAsync(block, blockId, blobAccessConditions, progressHandler, cancellationToken);
+
+                runningTasks.Add(task);
+                blocks.Add(blockId);
+
+                if (runningTasks.Count >= _threadCount)
+                {
+                    await Task.WhenAny(runningTasks).ConfigureAwait(false);
+
+                    runningTasks.RemoveAll(t => t.IsCompleted);
```

Comment:
Should we check for failure and stop uploading?  Or will the exception bubble up and break out of the Upload?

Reply:
Yes, we should observe all of them, I'll add a test.